### PR TITLE
Add the amp mixin

### PR DESCRIPTION
**Pull request by klobmathews**

### CODE REVIEW

**Comment by dave-shawley** [comment]
Manifest is incorrect.  I get the following when I try to install a generated source distribution:

```
s1033:sprockets.mixins.amqp% xxx/bin/pip install dist/sprockets.mixins.amqp-0.0.1.tar.gz
Processing ./dist/sprockets.mixins.amqp-0.0.1.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "/var/folders/48/hvbwyh4d2m5b78p_jqr4kjg0000231/T/pip-05p9pfl2-build/setup.py", line 25, in <module>
        install_requires=open('requires/installation.txt').read(),
    FileNotFoundError: [Errno 2] No such file or directory: 'requires/installation.txt'

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /var/folders/48/hvbwyh4d2m5b78p_jqr4kjg0000231/T/pip-05p9pfl2-build
```

Probably missing a `graft` statement in _MANIFEST.in_

**Comment by dave-shawley** [comment]
_sprockets/mixins/amqp/**init**.py_ shouldn't have a [namespace package declaration](https://www.python.org/dev/peps/pep-0420/#namespace-packages-today) in it unless you plan on installing additional packages under `sprockets.mixins.amqp`

File: setup.py [hunk 1]
```diff
@@ -2,47 +2,39 @@
...
-
-if sys.version_info < (3, 0):
-    tests_require.append('mock')
-
-setuptools.setup(
-    name='sprockets.amqp',
-    version=sprockets.amqp.__version__,
+setup(
+    name='sprockets.mixins.amqp',
     description='Making logs nicer since 2015!',
```
Comment (dave-shawley) [inline setup.py]: Looks surprisingly like `sprockets.logging`

File: sprockets/mixins/amqp/__init__.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+__import__('pkg_resources').declare_namespace(__name__)
+
+version_info = (0, 0, 1)
+__version__ = '.'.join(str(v) for v in version_info)
+
+__all__ = ('version_info', '__version__', 'RabbitMQRequestMixin')
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/__init__.py]: `RabbitMQRequestMixin` isn't an available symbol.  If you want to export a new symbol at the top-level so that a user can do `from sprockets.mixins.amqp import RabbitMQRequestMixin`, then you will need to import _amqp.py_ here and export it.  Just be careful since this module is imported by _setup.py_ when tornado is not installed.  IOW, do something like:

``` python
try:
   from sprockets.mixins.amqp.mixins import RabbitMQRequestMixin
except ImportError as err:
   def RabbitMQRequestMixin(*args, **kwargs):
      raise err
```

File: sprockets/mixins/amqp/amqp.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        username: tinman
+        password: tornado
+
+"""
+import os
+from tornado import gen, locks
+import logging
+import pika
+from pika.adapters import tornado_connection
+from tornado import web
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py]: NIT - ordering

File: sprockets/mixins/amqp/amqp.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        :param str app_id: Creating application id
+        :rtype: pika.BasicProperties
+
+        """
+        return pika.BasicProperties(content_type, content_encoding, headers,
+                                    delivery_mode, priority, correlation_id,
+                                    reply_to, expiration, message_id,
+                                    timestamp, message_type, user_id, app_id)
+
+    def _new_rabbitmq_connection(self):
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py]: Is there any value in having a one line function that simply calls another function?  Seems like 10 or so lines of fat to my eyes.

File: sprockets/mixins/amqp/amqp.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        global CONNECTION
+        LOGGER.info('Creating a new RabbitMQ connection')
+        CONNECTION = self._new_rabbitmq_connection()
+
+    def _new_message_properties(self, content_type=None, content_encoding=None,
+                                headers=None, delivery_mode=None,
+                                priority=None, correlation_id=None,
+                                reply_to=None, expiration=None,
+                                message_id=None, timestamp=None,
+                                message_type=None, user_id=None, app_id=None):
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py:58]: What is the value of this method?

File: sprockets/mixins/amqp/amqp.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+    def _new_rabbitmq_connection(self):
+        """Return a connection to RabbitMQ via the pika.Connection object.
+        When RabbitMQ is connected, on_rabbitmq_open will be called.
+
+        :rtype: pika.adapters.tornado_connection.TornadoConnection
+
+        """
+        return tornado_connection.TornadoConnection(self._rabbitmq_parameters,
+                                                    self.on_conn_open)
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py]: You should be handling the connection error case by specifying a callback as the [`on_open_error_callback` parameter](http://pika.readthedocs.org/en/0.10.0/modules/adapters/tornado.html)

File: sprockets/mixins/amqp/amqp.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+    @gen.coroutine
+    def prepare(self):
+        """Prepare the handler, ensuring RabbitMQ is connected or start a new
+        connection attempt.
+
+        Wait for the connection condition to notify that the rabbit connection
+        has been established.
+
+        This will timeout
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py]: Really?  But what happens when it does?

File: sprockets/mixins/amqp/amqp.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+        """
+        maybe_future = super(RabbitMQRequestMixin, self).prepare()
+        if maybe_future:
+            yield maybe_future
+
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py]: Should be checking for `self._finished` here since the call to `super().prepare()` may have finished the request.

File: sprockets/mixins/amqp/amqp.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    @gen.coroutine
+    def on_conn_open(self, connection):
+        """Called when RabbitMQ has been connected to.
+
+        :param pika.connection.Connection connection: The pika connection
+
+        """
+        global CONNECTION, CHANNEL
+        LOGGER.info('RabbitMQ is attempting to connect')
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py]: we've already connected, we are fetching a channel of communication at this point.

**Comment by dave-shawley** [comment]
What happens if the initial request is blocked waiting for the ["connected condition" in `prepare`](https://github.com/sprockets/sprockets.mixins.amqp/blob/set-up-amqp/sprockets/mixins/amqp/amqp.py#L203) and a new request comes in?  The `CONNECTION` instance is not `None` so `self._rabbitmq_is_closed` is `False` and `prepare` will pass without being connected to rabbit.  Might not matter since any use of `_publish_message` is going to fail.

Be aware that fixing this by not setting `CONNECTION` until the `on_conn_open` is called will cause you to potentially leak connection instances.  This is probably better modeled using a [`tornado.locks.Event`](http://www.tornadoweb.org/en/stable/locks.html#event) to track whether we are connected or not instead of a `Condition`.

File: sprockets/mixins/amqp/amqp.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+        """
+        maybe_future = super(AMQPMixin, self).prepare()
+        if maybe_future:
+            yield maybe_future
+
+        if self._finished:
+            return
+
+        if not self.connection and self.amqp_state != 'connecting':
```
Comment (dave-shawley) [inline sprockets/mixins/amqp/amqp.py]: AFAICT, `amqp_state` will always be one of `{'CONNECTED', 'CONNECTING', 'DISCONNECTED'}`.  Should this be `self.amqp_state != AMQP_STATES['connecting']`?  The same comment applies to line 208 below.  We can probably jettison the `AMQP_STATES` dict altogether since it really doesn't buy us anything.  That is, unless I am missing something.

**Comment by dave-shawley** [comment]
+1 from me, any objections @gmr ?

File: setup.py [hunk 11]
```diff
@@ -2,47 +2,39 @@
...
+    install_requires=open('requires/installation.txt').read(),
+    author='AWeber Communications, Inc.',
+    author_email='api@aweber.com',
+    url='https://github.com/sprockets/sprockets.mixins.amqp.git',
     classifiers=[
-        'Development Status :: 4 - Beta',
+        'Development Status :: 1 - Planning',
         'Intended Audience :: Developers',
-        'License :: OSI Approved :: BSD License',
+        'License :: Other/Proprietary License',
```
Comment (gmr) [inline setup.py]: What's the proprietary license?

File: setup.py [hunk 12]
```diff
@@ -2,47 +2,39 @@
...
         'Natural Language :: English',
         'Operating System :: OS Independent',
         'Programming Language :: Python :: 2',
         'Programming Language :: Python :: 2.7',
         'Programming Language :: Python :: 3',
         'Programming Language :: Python :: 3.4',
-        'Programming Language :: Python :: Implementation :: CPython',
-        'Programming Language :: Python :: Implementation :: PyPy',
```
Comment (gmr) [inline setup.py]: Is there a reason why this doesn't support CPython or PyPy?

File: sprockets/mixins/amqp/amqp.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+    be made on the next request.
+
+    Expects the :envvar:`AMQP` environment variable to construct
+    :class:`pika.connection.URLParameters`.
+
+    https://github.com/gmr/tinman/blob/master/tinman/handlers/rabbitmq.py
+
+    """
+    amqp_channel_ready = locks.Event()
+    amqp_ready = locks.Event()
```
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: Why have both `amqp_ready` and `amqp_channel_ready`?

File: setup.py [hunk 14]
```diff
@@ -2,47 +2,39 @@
...
+    description='Mixin for publishing events to RabbitMQ',
+    version=version,
+    packages=find_packages(),
+    namespace_packages=['sprockets', 'sprockets.mixins'],
+    test_suite='nose.collector',
     long_description=codecs.open('README.rst', encoding='utf-8').read(),
-    url='https://github.com/sprockets/sprockets.amqp.git',
-    author='Dave Shawley',
-    author_email='daves@aweber.com',
-    license='BSD',
```
Comment (gmr) [inline setup.py]: license is a important field for submitting to PyPi, if you're looking to change the license, you should specify what it is.

File: sprockets/mixins/amqp/amqp.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+        """Connect to RabbitMQ and assign a class attribute"""
+        LOGGER.info('Creating a new RabbitMQ connection')
+        custom_ioloop = ioloop.IOLoop.current()
+        AMQPMixin.connection = TornadoConnection(self._rabbit_params,
+                                                 self.on_conn_open,
+                                                 custom_ioloop=custom_ioloop)
+
+    @gen.coroutine
+    def _create_new_channel(self):
+        """"""
```
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: Weird doc string here.

File: sprockets/mixins/amqp/amqp.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+        """Called when RabbitMQ has been connected to.
+
+        :param pika.connection.Connection connection: The pika connection
+
+        """
+        LOGGER.info('RabbitMQ has created a connection')
+        AMQPMixin.connection = connection
+        AMQPMixin.connection.add_on_close_callback(self.on_conn_close)
+        yield self._create_new_channel()
+        self.amqp_ready.set()
```
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: Why do you need this?

File: sprockets/mixins/amqp/amqp.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+        """Called when the RabbitMQ accepts the channel close request.
+
+        :param pika.channel.Channel channel: The channel closed with RabbitMQ
+
+        """
+        LOGGER.info('Channel %i is closed for communication with RabbitMQ',
+                    channel.channel_number)
+        AMQPMixin.channel = None
+        self.amqp_channel_ready.clear()
+        AMQPMixin.amqp_state = AMQP_STATES['disconnected']
```
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: So the channel is closed, but then what? Why not open a new channel right here?

File: sprockets/mixins/amqp/amqp.py [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+
+    def reset_amqp_state(self):
+        """ Clear the amqp_ready event, clear the amqp_channel_ready event and
+        set the AMQP_STATE to connecting.
+
+        """
+        AMQPMixin.amqp_state = AMQP_STATES['connecting']
```
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: Is the default state "connecting"? Seems odd to have it connecting unless it is connecting

File: sprockets/mixins/amqp/amqp.py [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+            try:
+                self._connect_to_rabbitmq()
+                yield self.amqp_ready.wait(timeout=WAIT_TIMEOUT)
+            except gen.TimeoutError as e:
+                self.set_status(504)
+                self.finish('RabbitMQ connection timed out')
+                return
+
+        if not self.channel and self.amqp_state != AMQP_STATES['connecting']:
+            self.reset_amqp_state()
```
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: Seems weird to reset all of your events here

File: sprockets/mixins/amqp/amqp.py (hunk 13 above)
Comment (klobmathews) [inline sprockets/mixins/amqp/amqp.py]: Because you can't add an on_close_callback to the channel until it is established

File: sprockets/mixins/amqp/amqp.py (hunk 16 above)
Comment (klobmathews) [inline sprockets/mixins/amqp/amqp.py]: Because you can't add an on_close_callback to the channel until it is established

File: sprockets/mixins/amqp/amqp.py (hunk 13 above)
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: Right but you have a connection chain of:

Open Connection -> on Open of Connection -> Open Channel -> on Open of Channel

So why not just have a single event that is set when your "on Open of Channel" occurs. You don't need to track the state of connected but with no open channels. When a channel is closed, you can clear the event and open the channel again.

File: sprockets/mixins/amqp/amqp.py (hunk 16 above)
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: See my note above, I don't think you need the three things keeping track of state.

File: requires/installation.txt [hunk 20]
```diff
@@ -0,0 +1,5 @@
+pika>=0.9.14,<1.0
```
Comment (gmr) [inline requires/installation.txt]: pika>=0.10.0

File: setup.py [hunk 21]
```diff
@@ -2,30 +2,33 @@
...
-    extras_require={'tornado': ['tornado>3,<5']},
+    install_requires=open('requires/installation.txt').read(),
+    author='AWeber Communications, Inc.',
+    author_email='api@aweber.com',
+    license=codecs.open('LICENSE', encoding='utf-8').read(),
```
Comment (gmr) [inline setup.py]: This should just be the license name. Not doing so can break some badging and the name is enough.

File: setup.py [hunk 22]
```diff
@@ -2,30 +2,33 @@
...
+    url='https://github.com/sprockets/sprockets.mixins.amqp.git',
     classifiers=[
-        'Development Status :: 4 - Beta',
+        'Development Status :: 1 - Planning',
```
Comment (gmr) [inline setup.py]: Is it really `Planning` status? `Development Status :: 3 - Alpha` seems more appropriate.

File: setup.py [hunk 23]
```diff
@@ -36,13 +39,5 @@
         'Programming Language :: Python :: 3.4',
         'Programming Language :: Python :: Implementation :: CPython',
         'Programming Language :: Python :: Implementation :: PyPy',
-        'Topic :: Software Development :: Libraries',
```
Comment (gmr) [inline setup.py]: Why did you remove the other Topic classifiers?

File: sprockets/mixins/amqp/amqp.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+    @property
+    def _rabbit_params(self):
+        """Return a pika URLParameters object using the :envvar:`AMQP`
+        environment variable which identifies the Rabbit MQ server as a URL
+        ready for :class:`pika.connection.URLParameters`.
+
+        :rtype: pika.URLParameters
+
+        """
+        LOGGER.debug('URLParameters: %s',
```
Comment (gmr) [inline sprockets/mixins/amqp/amqp.py]: Won't this blow up with a `KeyError` if the environment variable is not set? And why create two objects, just for logging?  How about something like:

``` python
params = pika.URLParameters(os.environ.get('AMQP', 'amqp://guest:guest@localhost:5672/%2f'))
LOGGER.debug('RabbitMQ connection params: %r', params)
return params
```

File: setup.py (hunk 21 above)
Comment (klobmathews) [inline setup.py]: :+1:  If that's the case, we might want to fix these projects also:
https://github.com/sprockets/sprockets.mixins.statsd/blob/master/setup.py#L11
https://github.com/sprockets/sprockets.mixins.correlation/blob/master/setup.py#L36
https://github.com/sprockets/sprockets.mixins.postgresql/blob/master/setup.py#L38
https://github.com/sprockets/sprockets.mixins.redis/blob/master/setup.py#L38
https://github.com/sprockets/sprockets.mixins.json_error/blob/master/setup.py#L38
https://github.com/sprockets/sprockets.clients.cassandra/blob/master/setup.py#L35
https://github.com/sprockets/sprockets.clients.statsd/blob/master/setup.py#L38
https://github.com/sprockets/sprockets.clients.redis/blob/master/setup.py#L34
https://github.com/sprockets/sprockets.clients.postgresql/blob/master/setup.py#L14
https://github.com/sprockets/sprockets.clients.memcached/blob/master/setup.py#L23

File: setup.py (hunk 22 above)
Comment (klobmathews) [inline setup.py]: :+1:

File: sprockets/mixins/amqp/amqp.py (hunk 24 above)
Comment (klobmathews) [inline sprockets/mixins/amqp/amqp.py]: :+1:
